Why: The MPI stubs declare every message buffer as INTEGER BUF(*), while
 MUMPS passes buffers of every type and rank (and elements of pointer
 arrays). Without -fallow-argument-mismatch, which CRAN's -pedantic turns
 back into an error, gfortran >= 10 rejects those calls. mpif.h now gives
 the 18 buffer routines explicit interfaces whose buffers are type-, kind-
 and rank-agnostic (!GCC$ / !DEC$ ATTRIBUTES NO_ARG_CHECK, flang's
 !DIR$ IGNORE_TKR (TKRC)), as MPI implementations do. The constants move to
 mpif_const.h, which mpi.f includes instead (a routine cannot see an
 interface for itself). MUMPS_COPY, which passed its INTEGER buffers to
 copy routines of each data type, now copies INTEGER words.
Upstream: not sent; worth offering (fixes gfortran >= 10 builds of libseq
 without -fallow-argument-mismatch).
Obsolete when: upstream libseq has explicit buffer interfaces.

diff --git a/libseq/mpi.f b/libseq/mpi.f
index 1d268ce..33e78ec 100644
--- a/libseq/mpi.f
+++ b/libseq/mpi.f
@@ -24,7 +24,7 @@ C******************************************************************
       SUBROUTINE MPI_BSEND( BUF, CNT, DATATYPE, DEST, TAG, COMM,
      &            IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, DATATYPE, DEST, TAG, COMM, IERR
       INTEGER BUF(*)
       WRITE(*,*) 'Error. MPI_BSEND should not be called.'
@@ -35,7 +35,7 @@ C******************************************************************
 C***********************************************************************
       SUBROUTINE MPI_BUFFER_ATTACH(BUF, CNT,  IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, IERR
       INTEGER BUF(*)
       IERR = 0
@@ -44,7 +44,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_BUFFER_DETACH(BUF, CNT,  IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, IERR
       INTEGER BUF(*)
            IERR = 0
@@ -222,7 +222,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_BARRIER( COMM, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER COMM, IERR
       IERR = 0
       RETURN
@@ -239,7 +239,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_BCAST( BUFFER, CNT, DATATYPE, ROOT, COMM, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, DATATYPE, ROOT, COMM, IERR
       INTEGER BUFFER( * )
       IERR = 0
@@ -248,7 +248,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_CANCEL( IREQ, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER IREQ, IERR
       IERR = 0
       RETURN
@@ -256,7 +256,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_COMM_CREATE( COMM, GROUP, COMM2, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER COMM, GROUP, COMM2, IERR
       IERR = 0
       RETURN
@@ -264,7 +264,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_COMM_DUP( COMM, COMM2, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER COMM, COMM2, IERR
       IERR = 0
       RETURN
@@ -272,7 +272,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_COMM_FREE( COMM, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER COMM, IERR
       IERR = 0
       RETURN
@@ -280,7 +280,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_COMM_GROUP( COMM, GROUP, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER COMM, GROUP, IERR
       IERR = 0
       RETURN
@@ -288,7 +288,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_COMM_RANK( COMM, RANK, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER COMM, RANK, IERR
       RANK = 0
       IERR = 0
@@ -297,7 +297,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_COMM_SIZE( COMM, SIZE, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER COMM, SIZE, IERR
       SIZE = 1
       IERR = 0
@@ -306,7 +306,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_COMM_SPLIT( COMM, COLOR, KEY, COMM2, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER COMM, COLOR, KEY, COMM2, IERR
       IERR = 0
       RETURN
@@ -314,7 +314,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_FINALIZE( IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER IERR
       IERR = 0
       RETURN
@@ -322,7 +322,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_GET_COUNT( STATUS, DATATYPE, CNT, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER DATATYPE, CNT, IERR
       INTEGER STATUS( MPI_STATUS_SIZE )
       WRITE(*,*) 'Error. MPI_GET_CNT should not be called.'
@@ -333,7 +333,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_GROUP_FREE( GROUP, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER GROUP, IERR
       IERR = 0
       RETURN
@@ -341,7 +341,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_GROUP_RANGE_EXCL( GROUP, N, RANGES, GROUP2, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER GROUP, N, GROUP2, IERR
       INTEGER RANGES(*)
       IERR = 0
@@ -350,7 +350,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_GROUP_SIZE( GROUP, SIZE, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER GROUP, SIZE, IERR
       SIZE = 1 ! Or should it be zero ?
       IERR = 0
@@ -359,7 +359,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_INIT_THREAD(MPI_THREAD_REQ, THREAD_SUPPORT, IERR)
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER IERR, MPI_THREAD_REQ, THREAD_SUPPORT
       IERR = 0
       THREAD_SUPPORT=MPI_THREAD_REQ
@@ -368,7 +368,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_INIT(IERR)
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER IERR
       IERR = 0
       RETURN
@@ -376,7 +376,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_INITIALIZED( FLAG, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       LOGICAL FLAG
       INTEGER IERR
       FLAG = .TRUE.
@@ -386,7 +386,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_IPROBE( SOURCE, TAG, COMM, FLAG, STATUS, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER SOURCE, TAG, COMM, IERR
       INTEGER STATUS(MPI_STATUS_SIZE)
       LOGICAL FLAG
@@ -398,7 +398,7 @@ C***********************************************************************
       SUBROUTINE MPI_IRECV( BUF, CNT, DATATYPE, SOURCE, TAG, COMM,
      &           IREQ, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, DATATYPE, SOURCE, TAG, COMM, IREQ, IERR
       INTEGER BUF(*)
       IERR = 0
@@ -408,7 +408,7 @@ C***********************************************************************
       SUBROUTINE MPI_ISEND( BUF, CNT, DATATYPE, DEST, TAG, COMM,
      &           IREQ, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, DATATYPE, DEST, TAG, COMM, IERR, IREQ
       INTEGER BUF(*)
       WRITE(*,*) 'Error. MPI_ISEND should not be called.'
@@ -454,7 +454,7 @@ C***********************************************************************
       SUBROUTINE MPI_PACK( INBUF, INCNT, DATATYPE, OUTBUF, OUTCNT,
      &           POSITION, COMM, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER INCNT, DATATYPE, OUTCNT, POSITION, COMM, IERR
       INTEGER INBUF(*), OUTBUF(*)
       WRITE(*,*) 'Error. MPI_PACKED should not be called.'
@@ -465,7 +465,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_PACK_SIZE( INCNT, DATATYPE, COMM, SIZE, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER INCNT, DATATYPE, COMM, SIZE, IERR
       WRITE(*,*) 'Error. MPI_PACK_SIZE should not be called.'
       STOP
@@ -475,7 +475,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_PROBE( SOURCE, TAG, COMM, STATUS, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER SOURCE, TAG, COMM, IERR
       INTEGER STATUS( MPI_STATUS_SIZE )
       WRITE(*,*) 'Error. MPI_PROBE should not be called.'
@@ -487,7 +487,7 @@ C***********************************************************************
       SUBROUTINE MPI_RECV( BUF, CNT, DATATYPE, SOURCE, TAG, COMM,
      &           STATUS, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, DATATYPE, SOURCE, TAG, COMM, IERR
       INTEGER BUF(*), STATUS(MPI_STATUS_SIZE)
       WRITE(*,*) 'Error. MPI_RECV should not be called.'
@@ -498,7 +498,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_REQUEST_FREE( IREQ, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER IREQ, IERR
       IERR = 0
       RETURN
@@ -506,7 +506,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_SEND( BUF, CNT, DATATYPE, DEST, TAG, COMM, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, DATATYPE, DEST, TAG, COMM, IERR
       INTEGER BUF(*)
       WRITE(*,*) 'Error. MPI_SEND should not be called.'
@@ -517,7 +517,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_SSEND( BUF, CNT, DATATYPE, DEST, TAG, COMM, IERR)
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, DATATYPE, DEST, TAG, COMM, IERR
       INTEGER BUF(*)
       WRITE(*,*) 'Error. MPI_SSEND should not be called.'
@@ -528,7 +528,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_TEST( IREQ, FLAG, STATUS, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER IREQ, IERR
       INTEGER STATUS( MPI_STATUS_SIZE )
       LOGICAL FLAG
@@ -539,7 +539,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_TESTANY( CNT, IREQARRAY, IND, FLAG, STATUS, IERR)
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       LOGICAL FLAG
       INTEGER IND, IERR, CNT
       INTEGER IREQARRAY(IND), STATUS(MPI_STATUS_SIZE)
@@ -551,7 +551,7 @@ C***********************************************************************
       SUBROUTINE MPI_UNPACK( INBUF, INSIZE, POSITION, OUTBUF, OUTCNT,
      &           DATATYPE, COMM, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER INSIZE, POSITION, OUTCNT, DATATYPE, COMM, IERR
       INTEGER INBUF(*), OUTBUF(*)
       WRITE(*,*) 'Error. MPI_UNPACK should not be called.'
@@ -562,7 +562,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_WAIT( IREQ, STATUS, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER IREQ, IERR
       INTEGER STATUS( MPI_STATUS_SIZE )
       WRITE(*,*) 'Error. MPI_WAIT should not be called.'
@@ -573,7 +573,7 @@ C***********************************************************************
 C***********************************************************************
       SUBROUTINE MPI_WAITALL( CNT, ARRAY_OF_REQUESTS, STATUS, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, IERR
       INTEGER STATUS( MPI_STATUS_SIZE )
       INTEGER ARRAY_OF_REQUESTS( CNT )
@@ -586,7 +586,7 @@ C***********************************************************************
       SUBROUTINE MPI_WAITANY( CNT, ARRAY_OF_REQUESTS, INDEX, STATUS,
      &           IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER CNT, INDEX, IERR
       INTEGER STATUS( MPI_STATUS_SIZE )
       INTEGER ARRAY_OF_REQUESTS( CNT )
@@ -624,7 +624,7 @@ C***********************************************************************
      &     INFO, COMM, CPTR_ARRAY, WIN, IERR )
       USE, INTRINSIC :: ISO_C_BINDING, ONLY : C_PTR
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER :: DISP_UNIT, INFO, COMM, WIN, IERR
       INTEGER(KIND=MPI_ADDRESS_KIND) :: SIZE_ARRAY_BYTES
       TYPE(C_PTR) :: CPTR_ARRAY
@@ -637,7 +637,7 @@ C***********************************************************************
      &     DISP_UNIT, CPTR_ARRAY, IERR )
       USE, INTRINSIC :: ISO_C_BINDING, ONLY : C_PTR
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER :: WIN, RANK, DISP_UNIT, IERR
       INTEGER(KIND=MPI_ADDRESS_KIND) :: SIZE_ARRAY_BYTES
       TYPE(C_PTR) :: CPTR_ARRAY
@@ -663,33 +663,46 @@ C***********************************************************************
       SUBROUTINE MUMPS_COPY( CNT, SENDBUF, RECVBUF, SS, RS,
      &                       DATATYPE, IERR )
       IMPLICIT NONE
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
 C     SS/RS: shift for SENDBUF/RECVBUF
       INTEGER CNT, DATATYPE, IERR, SS, RS
       INTEGER SENDBUF(*), RECVBUF(*)
+C     rmumps: the buffers are INTEGER here whatever the data type, so
+C     instead of passing them to copy routines of the data type (an
+C     argument mismatch), copy W INTEGER words per element, where W is
+C     the size of the data type in INTEGER storage units.
+      INTEGER W, IS
+      LOGICAL L
+      REAL R4
+      DOUBLE PRECISION R8
+      COMPLEX C4
+      COMPLEX(kind=kind(0.0D0)) C8
+      INTEGER(8) I8
+      IS = STORAGE_SIZE(SENDBUF(1))
       IF ( DATATYPE .EQ. MPI_INTEGER ) THEN
-      CALL MUMPS_COPY_INTEGER( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = 1
       ELSEIF ( DATATYPE .EQ. MPI_LOGICAL ) THEN
-      CALL MUMPS_COPY_LOGICAL( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = STORAGE_SIZE(L) / IS
       ELSE IF ( DATATYPE .EQ. MPI_REAL ) THEN
-      CALL MUMPS_COPY_REAL( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = STORAGE_SIZE(R4) / IS
       ELSE IF ( DATATYPE .EQ. MPI_DOUBLE_PRECISION .OR.
      &        DATATYPE .EQ. MPI_REAL8 ) THEN
-      CALL MUMPS_COPY_DOUBLE_PRECISION( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = STORAGE_SIZE(R8) / IS
       ELSE IF ( DATATYPE .EQ. MPI_COMPLEX ) THEN
-      CALL MUMPS_COPY_COMPLEX( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = STORAGE_SIZE(C4) / IS
       ELSE IF ( DATATYPE .EQ. MPI_DOUBLE_COMPLEX ) THEN
-      CALL MUMPS_COPY_DOUBLE_COMPLEX( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = STORAGE_SIZE(C8) / IS
       ELSE IF ( DATATYPE .EQ. MPI_2DOUBLE_PRECISION) THEN
-      CALL MUMPS_COPY_2DOUBLE_PRECISION( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = 2 * STORAGE_SIZE(R8) / IS
       ELSE IF ( DATATYPE .EQ. MPI_2INTEGER) THEN
-      CALL MUMPS_COPY_2INTEGER( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = 2
       ELSE IF ( DATATYPE .EQ. MPI_INTEGER8) THEN
-      CALL MUMPS_COPY_INTEGER8( SENDBUF, RECVBUF, CNT, SS, RS )
+        W = STORAGE_SIZE(I8) / IS
       ELSE
         IERR=1
         RETURN
       END IF
+      CALL MUMPS_COPY_INTEGER( SENDBUF, RECVBUF, CNT*W, SS*W, RS*W )
       IERR=0
       RETURN
       END SUBROUTINE MUMPS_COPY
@@ -787,7 +800,7 @@ C     DOUBLE COMPLEX S(N),R(N)
       END
       LOGICAL FUNCTION MUMPS_IS_IN_PLACE( SENDBUF, CNT )
       INTEGER SENDBUF(*), CNT
-      INCLUDE 'mpif.h'
+      INCLUDE 'mpif_const.h'
       INTEGER(8) :: I
 C     Check address using C code
       MUMPS_IS_IN_PLACE = .FALSE.
diff --git a/libseq/mpif.h b/libseq/mpif.h
index 9b3bc55..0cb3432 100644
--- a/libseq/mpif.h
+++ b/libseq/mpif.h
@@ -7,106 +7,162 @@
 !  Mumps Technologies, University of Bordeaux.
 !
 !  This version of MUMPS is provided to you free of charge. It is
-!  released under the CeCILL-C license 
+!  released under the CeCILL-C license
 !  (see doc/CeCILL-C_V1-en.txt, doc/CeCILL-C_V1-fr.txt, and
 !  https://cecill.info/licences/Licence_CeCILL-C_V1-en.html)
 !
 !
 !     Stub mpif.h file including symbols used by MUMPS.
 !
-      INTEGER MPI_2DOUBLE_PRECISION
-      INTEGER MPI_2INTEGER
-      INTEGER MPI_2REAL
-      INTEGER MPI_ANY_SOURCE
-      INTEGER MPI_ANY_TAG
-      INTEGER MPI_BYTE
-      INTEGER MPI_CHARACTER
-      INTEGER MPI_COMM_NULL
-      INTEGER MPI_COMM_WORLD
-      INTEGER MPI_COMPLEX
-      INTEGER MPI_DOUBLE_COMPLEX
-      INTEGER MPI_DOUBLE_PRECISION
-      INTEGER MPI_INTEGER
-      INTEGER MPI_LOGICAL
-      INTEGER MPI_MAX
-      INTEGER MPI_MAX_PROCESSOR_NAME
-      INTEGER MPI_MAXLOC
-      INTEGER MPI_MIN
-      INTEGER MPI_MINLOC
-      INTEGER MPI_PACKED
-      INTEGER MPI_PROD
-      INTEGER MPI_REAL
-      INTEGER MPI_REPLACE
-      INTEGER MPI_REQUEST_NULL
-      INTEGER MPI_SOURCE
-      INTEGER MPI_STATUS_SIZE
-      INTEGER MPI_SUM
-      INTEGER MPI_TAG
-      INTEGER MPI_UNDEFINED
-      INTEGER MPI_WTIME_IS_GLOBAL
-      INTEGER MPI_LOR
-      INTEGER MPI_LAND
-      INTEGER MPI_INTEGER8
-      INTEGER MPI_REAL8
-      INTEGER MPI_COMM_SELF
-      INTEGER MPI_BSEND_OVERHEAD
-      INTEGER MPI_THREAD_SINGLE
-      INTEGER MPI_THREAD_FUNNELED
-      INTEGER MPI_THREAD_MULTIPLE
-      INTEGER MPI_BOR
-      INTEGER MPI_WIN_NULL
-      INTEGER MPI_ADDRESS_KIND
-      INTEGER MPI_INFO_NULL
-      INTEGER MPI_SUCCESS
-      INTEGER MPI_ERRORS_RETURN
-      PARAMETER (MPI_2DOUBLE_PRECISION=1)
-      PARAMETER (MPI_2INTEGER=2)
-      PARAMETER (MPI_2REAL=3)
-      PARAMETER (MPI_ANY_SOURCE=4)
-      PARAMETER (MPI_ANY_TAG=5)
-      PARAMETER (MPI_BYTE=6)
-      PARAMETER (MPI_CHARACTER=7)
-      PARAMETER (MPI_COMM_NULL=8)
-      PARAMETER (MPI_COMM_WORLD=9)
-      PARAMETER (MPI_COMPLEX=10)
-      PARAMETER (MPI_DOUBLE_COMPLEX=11)
-      PARAMETER (MPI_DOUBLE_PRECISION=12)
-      PARAMETER (MPI_INTEGER=13)
-      PARAMETER (MPI_LOGICAL=14)
-      PARAMETER (MPI_MAX=15)
-      PARAMETER (MPI_MAX_PROCESSOR_NAME=31)
-      PARAMETER (MPI_MAXLOC=16)
-      PARAMETER (MPI_MIN=17)
-      PARAMETER (MPI_MINLOC=18)
-      PARAMETER (MPI_PACKED=19)
-      PARAMETER (MPI_PROD=20)
-      PARAMETER (MPI_REAL=21)
-      PARAMETER (MPI_REPLACE=22)
-      PARAMETER (MPI_REQUEST_NULL=23)
-      PARAMETER (MPI_SOURCE=1)
-      PARAMETER (MPI_STATUS_SIZE=2)
-      PARAMETER (MPI_SUM=26)
-      PARAMETER (MPI_TAG=2)
-      PARAMETER (MPI_UNDEFINED=28)
-      PARAMETER (MPI_WTIME_IS_GLOBAL=30)
-      PARAMETER (MPI_LOR=31)
-      PARAMETER (MPI_LAND=32)
-      PARAMETER (MPI_INTEGER8=33)
-      PARAMETER (MPI_REAL8=34)
-      PARAMETER (MPI_COMM_SELF=35)
-      PARAMETER (MPI_THREAD_SINGLE=36)
-      PARAMETER (MPI_THREAD_FUNNELED=37)
-      PARAMETER (MPI_THREAD_MULTIPLE=38)
-      PARAMETER (MPI_BOR=39)
-      PARAMETER (MPI_WIN_NULL=40)
-      PARAMETER (MPI_ADDRESS_KIND=8)
-      PARAMETER (MPI_INFO_NULL=41)
-      PARAMETER (MPI_SUCCESS=0)
-      PARAMETER (MPI_ERRORS_RETURN=42)
-
-      INTEGER MPI_IN_PLACE
-      COMMON/mpif_libseq/MPI_IN_PLACE
-
-      PARAMETER (MPI_BSEND_OVERHEAD=0)
-      DOUBLE PRECISION MPI_WTIME
-      EXTERNAL MPI_WTIME
+!     rmumps: the constants moved to mpif_const.h (which mpi.f includes
+!     directly); this file adds explicit interfaces for the routines that
+!     take message buffers. Their buffers are declared type-, kind- and
+!     rank-agnostic for gfortran (NO_ARG_CHECK), Intel (NO_ARG_CHECK) and
+!     LLVM flang (IGNORE_TKR), as in the mpif.h of MPI implementations,
+!     so callers may pass any data without argument-mismatch errors.
+!
+      INCLUDE 'mpif_const.h'
+      INTERFACE
+      SUBROUTINE MPI_ALLREDUCE( SENDBUF, RECVBUF, CNT, DATATYPE,
+     &                          OPERATION, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DIR$ IGNORE_TKR (TKRC) SENDBUF, RECVBUF
+      INTEGER SENDBUF(*), RECVBUF(*)
+      INTEGER CNT, DATATYPE, OPERATION, COMM, IERR
+      END SUBROUTINE MPI_ALLREDUCE
+      SUBROUTINE MPI_ALLTOALL( SENDBUF, SENDCNT, SENDTYPE,
+     &                         RECVBUF, RECVCNT, RECVTYPE, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DIR$ IGNORE_TKR (TKRC) SENDBUF, RECVBUF
+      INTEGER SENDBUF(*), RECVBUF(*)
+      INTEGER SENDCNT, SENDTYPE, RECVCNT, RECVTYPE, COMM, IERR
+      END SUBROUTINE MPI_ALLTOALL
+      SUBROUTINE MPI_ALLTOALLV( SENDBUF, SENDCNT, SDISPLS, SENDTYPE,
+     &                         RECVBUF, RECVCNT, RDISPLS, RECVTYPE,
+     &                         COMM, IERR)
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DIR$ IGNORE_TKR (TKRC) SENDBUF, RECVBUF
+      INTEGER SENDBUF(*), RECVBUF(*)
+      INTEGER SENDCNT, SENDTYPE, RECVCNT, RECVTYPE, COMM, IERR
+      INTEGER SDISPLS(*), RDISPLS(*)
+      END SUBROUTINE MPI_ALLTOALLV
+      SUBROUTINE MPI_BCAST( BUFFER, CNT, DATATYPE, ROOT, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUFFER
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUFFER
+!DIR$ IGNORE_TKR (TKRC) BUFFER
+      INTEGER BUFFER(*)
+      INTEGER CNT, DATATYPE, ROOT, COMM, IERR
+      END SUBROUTINE MPI_BCAST
+      SUBROUTINE MPI_BSEND( BUF, CNT, DATATYPE, DEST, TAG, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DIR$ IGNORE_TKR (TKRC) BUF
+      INTEGER BUF(*)
+      INTEGER CNT, DATATYPE, DEST, TAG, COMM, IERR
+      END SUBROUTINE MPI_BSEND
+      SUBROUTINE MPI_BUFFER_ATTACH( BUF, CNT, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DIR$ IGNORE_TKR (TKRC) BUF
+      INTEGER BUF(*)
+      INTEGER CNT, IERR
+      END SUBROUTINE MPI_BUFFER_ATTACH
+      SUBROUTINE MPI_BUFFER_DETACH( BUF, CNT, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DIR$ IGNORE_TKR (TKRC) BUF
+      INTEGER BUF(*)
+      INTEGER CNT, IERR
+      END SUBROUTINE MPI_BUFFER_DETACH
+      SUBROUTINE MPI_GATHER( SENDBUF, CNT, DATATYPE, RECVBUF, RECVCNT,
+     &                       RECTYPE, ROOT, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DIR$ IGNORE_TKR (TKRC) SENDBUF, RECVBUF
+      INTEGER SENDBUF(*), RECVBUF(*)
+      INTEGER CNT, DATATYPE, RECVCNT, RECTYPE, ROOT, COMM, IERR
+      END SUBROUTINE MPI_GATHER
+      SUBROUTINE MPI_GATHERV( SENDBUF, CNT, DATATYPE, RECVBUF, RECVCNT,
+     &                        DISPLS, RECTYPE, ROOT, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DIR$ IGNORE_TKR (TKRC) SENDBUF, RECVBUF
+      INTEGER SENDBUF(*), RECVBUF(*)
+      INTEGER CNT, DATATYPE, RECTYPE, ROOT, COMM, IERR
+      INTEGER RECVCNT(*), DISPLS(*)
+      END SUBROUTINE MPI_GATHERV
+      SUBROUTINE MPI_IRECV( BUF, CNT, DATATYPE, SOURCE, TAG, COMM,
+     &                      IREQ, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DIR$ IGNORE_TKR (TKRC) BUF
+      INTEGER BUF(*)
+      INTEGER CNT, DATATYPE, SOURCE, TAG, COMM, IREQ, IERR
+      END SUBROUTINE MPI_IRECV
+      SUBROUTINE MPI_ISEND( BUF, CNT, DATATYPE, DEST, TAG, COMM,
+     &                      IREQ, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DIR$ IGNORE_TKR (TKRC) BUF
+      INTEGER BUF(*)
+      INTEGER CNT, DATATYPE, DEST, TAG, COMM, IREQ, IERR
+      END SUBROUTINE MPI_ISEND
+      SUBROUTINE MPI_PACK( INBUF, INCNT, DATATYPE, OUTBUF, OUTCNT,
+     &                     POSITION, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: INBUF, OUTBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: INBUF, OUTBUF
+!DIR$ IGNORE_TKR (TKRC) INBUF, OUTBUF
+      INTEGER INBUF(*), OUTBUF(*)
+      INTEGER INCNT, DATATYPE, OUTCNT, POSITION, COMM, IERR
+      END SUBROUTINE MPI_PACK
+      SUBROUTINE MPI_RECV( BUF, CNT, DATATYPE, SOURCE, TAG, COMM,
+     &                     STATUS, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DIR$ IGNORE_TKR (TKRC) BUF
+      INTEGER BUF(*)
+      INTEGER CNT, DATATYPE, SOURCE, TAG, COMM, IERR
+      INTEGER STATUS(*)
+      END SUBROUTINE MPI_RECV
+      SUBROUTINE MPI_REDUCE( SENDBUF, RECVBUF, CNT, DATATYPE, OP,
+     &                       ROOT, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DIR$ IGNORE_TKR (TKRC) SENDBUF, RECVBUF
+      INTEGER SENDBUF(*), RECVBUF(*)
+      INTEGER CNT, DATATYPE, OP, ROOT, COMM, IERR
+      END SUBROUTINE MPI_REDUCE
+      SUBROUTINE MPI_REDUCE_SCATTER( SENDBUF, RECVBUF, RECVCNT,
+     &                               DATATYPE, OP, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: SENDBUF, RECVBUF
+!DIR$ IGNORE_TKR (TKRC) SENDBUF, RECVBUF
+      INTEGER SENDBUF(*), RECVBUF(*)
+      INTEGER RECVCNT, DATATYPE, OP, COMM, IERR
+      END SUBROUTINE MPI_REDUCE_SCATTER
+      SUBROUTINE MPI_SEND( BUF, CNT, DATATYPE, DEST, TAG, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DIR$ IGNORE_TKR (TKRC) BUF
+      INTEGER BUF(*)
+      INTEGER CNT, DATATYPE, DEST, TAG, COMM, IERR
+      END SUBROUTINE MPI_SEND
+      SUBROUTINE MPI_SSEND( BUF, CNT, DATATYPE, DEST, TAG, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: BUF
+!DIR$ IGNORE_TKR (TKRC) BUF
+      INTEGER BUF(*)
+      INTEGER CNT, DATATYPE, DEST, TAG, COMM, IERR
+      END SUBROUTINE MPI_SSEND
+      SUBROUTINE MPI_UNPACK( INBUF, INSIZE, POSITION, OUTBUF, OUTCNT,
+     &                       DATATYPE, COMM, IERR )
+!GCC$ ATTRIBUTES NO_ARG_CHECK :: INBUF, OUTBUF
+!DEC$ ATTRIBUTES NO_ARG_CHECK :: INBUF, OUTBUF
+!DIR$ IGNORE_TKR (TKRC) INBUF, OUTBUF
+      INTEGER INBUF(*), OUTBUF(*)
+      INTEGER INSIZE, POSITION, OUTCNT, DATATYPE, COMM, IERR
+      END SUBROUTINE MPI_UNPACK
+      END INTERFACE
diff --git b/libseq/mpif_const.h b/libseq/mpif_const.h
new file mode 100644
index 0000000..9b3bc55
--- /dev/null
+++ b/libseq/mpif_const.h
@@ -0,0 +1,112 @@
+!
+!  This file is part of MUMPS 5.9.1, released
+!  on Mon Jul 20 09:00:43 UTC 2026
+!
+!
+!  Copyright 1991-2026 CERFACS, CNRS, ENS Lyon, INP Toulouse, Inria,
+!  Mumps Technologies, University of Bordeaux.
+!
+!  This version of MUMPS is provided to you free of charge. It is
+!  released under the CeCILL-C license 
+!  (see doc/CeCILL-C_V1-en.txt, doc/CeCILL-C_V1-fr.txt, and
+!  https://cecill.info/licences/Licence_CeCILL-C_V1-en.html)
+!
+!
+!     Stub mpif.h file including symbols used by MUMPS.
+!
+      INTEGER MPI_2DOUBLE_PRECISION
+      INTEGER MPI_2INTEGER
+      INTEGER MPI_2REAL
+      INTEGER MPI_ANY_SOURCE
+      INTEGER MPI_ANY_TAG
+      INTEGER MPI_BYTE
+      INTEGER MPI_CHARACTER
+      INTEGER MPI_COMM_NULL
+      INTEGER MPI_COMM_WORLD
+      INTEGER MPI_COMPLEX
+      INTEGER MPI_DOUBLE_COMPLEX
+      INTEGER MPI_DOUBLE_PRECISION
+      INTEGER MPI_INTEGER
+      INTEGER MPI_LOGICAL
+      INTEGER MPI_MAX
+      INTEGER MPI_MAX_PROCESSOR_NAME
+      INTEGER MPI_MAXLOC
+      INTEGER MPI_MIN
+      INTEGER MPI_MINLOC
+      INTEGER MPI_PACKED
+      INTEGER MPI_PROD
+      INTEGER MPI_REAL
+      INTEGER MPI_REPLACE
+      INTEGER MPI_REQUEST_NULL
+      INTEGER MPI_SOURCE
+      INTEGER MPI_STATUS_SIZE
+      INTEGER MPI_SUM
+      INTEGER MPI_TAG
+      INTEGER MPI_UNDEFINED
+      INTEGER MPI_WTIME_IS_GLOBAL
+      INTEGER MPI_LOR
+      INTEGER MPI_LAND
+      INTEGER MPI_INTEGER8
+      INTEGER MPI_REAL8
+      INTEGER MPI_COMM_SELF
+      INTEGER MPI_BSEND_OVERHEAD
+      INTEGER MPI_THREAD_SINGLE
+      INTEGER MPI_THREAD_FUNNELED
+      INTEGER MPI_THREAD_MULTIPLE
+      INTEGER MPI_BOR
+      INTEGER MPI_WIN_NULL
+      INTEGER MPI_ADDRESS_KIND
+      INTEGER MPI_INFO_NULL
+      INTEGER MPI_SUCCESS
+      INTEGER MPI_ERRORS_RETURN
+      PARAMETER (MPI_2DOUBLE_PRECISION=1)
+      PARAMETER (MPI_2INTEGER=2)
+      PARAMETER (MPI_2REAL=3)
+      PARAMETER (MPI_ANY_SOURCE=4)
+      PARAMETER (MPI_ANY_TAG=5)
+      PARAMETER (MPI_BYTE=6)
+      PARAMETER (MPI_CHARACTER=7)
+      PARAMETER (MPI_COMM_NULL=8)
+      PARAMETER (MPI_COMM_WORLD=9)
+      PARAMETER (MPI_COMPLEX=10)
+      PARAMETER (MPI_DOUBLE_COMPLEX=11)
+      PARAMETER (MPI_DOUBLE_PRECISION=12)
+      PARAMETER (MPI_INTEGER=13)
+      PARAMETER (MPI_LOGICAL=14)
+      PARAMETER (MPI_MAX=15)
+      PARAMETER (MPI_MAX_PROCESSOR_NAME=31)
+      PARAMETER (MPI_MAXLOC=16)
+      PARAMETER (MPI_MIN=17)
+      PARAMETER (MPI_MINLOC=18)
+      PARAMETER (MPI_PACKED=19)
+      PARAMETER (MPI_PROD=20)
+      PARAMETER (MPI_REAL=21)
+      PARAMETER (MPI_REPLACE=22)
+      PARAMETER (MPI_REQUEST_NULL=23)
+      PARAMETER (MPI_SOURCE=1)
+      PARAMETER (MPI_STATUS_SIZE=2)
+      PARAMETER (MPI_SUM=26)
+      PARAMETER (MPI_TAG=2)
+      PARAMETER (MPI_UNDEFINED=28)
+      PARAMETER (MPI_WTIME_IS_GLOBAL=30)
+      PARAMETER (MPI_LOR=31)
+      PARAMETER (MPI_LAND=32)
+      PARAMETER (MPI_INTEGER8=33)
+      PARAMETER (MPI_REAL8=34)
+      PARAMETER (MPI_COMM_SELF=35)
+      PARAMETER (MPI_THREAD_SINGLE=36)
+      PARAMETER (MPI_THREAD_FUNNELED=37)
+      PARAMETER (MPI_THREAD_MULTIPLE=38)
+      PARAMETER (MPI_BOR=39)
+      PARAMETER (MPI_WIN_NULL=40)
+      PARAMETER (MPI_ADDRESS_KIND=8)
+      PARAMETER (MPI_INFO_NULL=41)
+      PARAMETER (MPI_SUCCESS=0)
+      PARAMETER (MPI_ERRORS_RETURN=42)
+
+      INTEGER MPI_IN_PLACE
+      COMMON/mpif_libseq/MPI_IN_PLACE
+
+      PARAMETER (MPI_BSEND_OVERHEAD=0)
+      DOUBLE PRECISION MPI_WTIME
+      EXTERNAL MPI_WTIME
